### PR TITLE
[Tapioca Add-on] Generate DSL RBIs in add-on mode

### PR DESCRIPTION
**Pull request by KaanOzkan**
### Motivation

Extracting from https://github.com/Shopify/tapioca/compare/main...tapioca-addon-feature-branch.
### Tests

Will be added later.

### CODE REVIEW

File: lib/tapioca/dsl/compiler.rb [hunk 1]
```diff
@@ -51,6 +51,13 @@ def requested_constants=(constants)
           @@requested_constants = constants # rubocop:disable Style/ClassVars
         end
 
+        sig { void }
+        def reset_state
+          @processable_constants = nil
+          @all_classes = nil
+          @all_modules = nil
+        end
```
Comment (KaanOzkan) [inline lib/tapioca/dsl/compiler.rb:59]: These are memoized but in LSP mode we want them to be calculated every time.

**Review by andyw8 — APPROVED** [review]

File: lib/ruby_lsp/tapioca/addon.rb [hunk 2]
```diff
@@ -63,7 +69,41 @@ def workspace_did_change_watched_files(changes)
         return unless T.must(@global_state).enabled_feature?(:tapiocaAddon)
         return unless @rails_runner_client # Client is not ready
 
-        nil
+        constants = changes.flat_map do |change|
+          path = URI(change[:uri]).to_standardized_path
```
Comment (andyw8) [inline lib/ruby_lsp/tapioca/addon.rb:73]: Perhaps extract the body of this to its own method for readability?

File: lib/ruby_lsp/tapioca/addon.rb [hunk 3]
```diff
@@ -63,7 +69,41 @@ def workspace_did_change_watched_files(changes)
...
+          next if path.end_with?("_test.rb", "_spec.rb")
+
+          case change[:type]
+          when Constant::FileChangeType::CREATED, Constant::FileChangeType::CHANGED
+            content = File.read(path)
+            current_checksum = Zlib.crc32(content).to_s
+
+            if change[:type] == Constant::FileChangeType::CHANGED && @file_checksums[path] == current_checksum
+              $stderr.puts "File has not changed. Skipping #{path}"
```
[the remark this answers is not in the document]
Comment (KaanOzkan) [reply lib/ruby_lsp/tapioca/addon.rb:82]: Looks like it's not part of the ruby API yet https://github.com/mtsmfm/language_server-protocol-ruby/blob/15d4388e84c65d9900d9dc1ca378e0073231d44b/lib/language_server/protocol/constant/message_type.rb#L4

File: lib/ruby_lsp/tapioca/addon.rb (hunk 2 above)
Comment (andyw8) [reply lib/ruby_lsp/tapioca/addon.rb:78]: I'm a little uncomfortable that `file_updated?` has side-effects. I'll take a look to see if there is another way we could structure it.

File: lib/ruby_lsp/tapioca/addon.rb [hunk 4]
```diff
@@ -23,24 +25,33 @@ def initialize
...
           # thread
           addon = T.cast(::RubyLsp::Addon.get("Ruby LSP Rails", ">= 0.3.17", "< 0.4"), ::RubyLsp::Rails::Addon)
           @rails_runner_client = addon.rails_runner_client
-          outgoing_queue << Notification.window_log_message("Activating Tapioca add-on v#{version}")
+          @outgoing_queue << Notification.window_log_message("Activating Tapioca add-on v#{version}")
           @rails_runner_client.register_server_addon(File.expand_path("server_addon.rb", __dir__))
         rescue IncompatibleApiError
           # The requested version for the Rails add-on no longer matches. We need to upgrade and fix the breaking
           # changes
+          outgoing_queue << Notification.window_log_message(
```
Comment (alexcrocha) [inline lib/ruby_lsp/tapioca/addon.rb:51]: Should this also be an instance variable?

```suggestion
          @outgoing_queue << Notification.window_log_message(
```

File: lib/tapioca/commands/abstract_dsl.rb [hunk 5]
```diff
@@ -77,6 +77,10 @@ def initialize(
 
       sig { params(outpath: Pathname, quiet: T::Boolean).returns(T::Set[Pathname]) }
       def generate_dsl_rbi_files(outpath, quiet:)
+        if @lsp_addon
+          pipeline.active_compilers.each(&:reset_state)
```
[the remark this answers is not in the document]
Comment (KaanOzkan) [reply lib/tapioca/commands/abstract_dsl.rb:81]: I think main issue is another method checking for `lsp_addon`. I don't want this code to be reusable for other clients. It's a unique thing (kind of hacky and not ideal 😬 ) for addon support.

**Pull request merged**